### PR TITLE
Add type hint to the automl module

### PR DESCRIPTION
**Pull request by smurf-1119**
## Description

Add type hint to the automl module `automl.metrics` and `automl.auto_estimator`

### How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 1]
```diff
@@ -53,14 +59,14 @@ def __init__(self,
...
     def from_torch(*,
-                   model_creator,
-                   optimizer,
-                   loss,
-                   logs_dir="/tmp/auto_estimator_logs",
-                   resources_per_trial=None,
-                   name="auto_pytorch_estimator",
-                   remote_dir=None,
-                   ):
+                   model_creator: function,
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:62]: Callable

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 2]
```diff
@@ -124,21 +130,22 @@ def from_keras(*,
...
-            n_sampling=1,
-            search_space=None,
-            search_alg=None,
-            search_alg_params=None,
-            scheduler=None,
-            scheduler_params=None,
-            feature_cols=None,
-            label_cols=None,
-            ):
+            data: Union[Callable, Tuple[ndarray, ndarray]],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:133]: and Dataloader as comments ray?

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 3]
```diff
@@ -124,21 +130,22 @@ def from_keras(*,
...
+            epochs: int=1,
+            validation_data: Optional[Union[Callable, Tuple[ndarray, ndarray]]]=None,
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:135]: same as data

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 4]
```diff
@@ -124,21 +130,22 @@ def from_keras(*,
...
+            metric: Optional[Union[Callable, str]]=None,
+            metric_mode: Optional[str]=None,
+            metric_threshold: Optional[float]=None,
+            n_sampling: int=1,
+            search_space: Optional[Dict[str, Union[Float, Categorical,
+                                                   Dict[str, List[bool]]]]]=None,
+            search_alg: Optional[str]=None,
+            search_alg_params: None=None,
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:143]: Should not be None

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 5]
```diff
@@ -124,21 +130,22 @@ def from_keras(*,
...
+            scheduler: Optional[str]=None,
+            scheduler_params: None=None,
+            feature_cols: Optional[DataFrame]=None,
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:146]: Optional[List[str]]

**Review by leonardozcm — COMMENTED** [review]
@smurf-1119 May test this with mypy?
https://github.com/intel-analytics/BigDL/pull/6969/files#diff-f0eb7bb6f684bfeb2c36e89a05d9c8c4fc15e3c62eb1c80e44d20c4f2cd1bdabR18-R22

**Comment by liangs6212** [comment]
Could we ignore the check for "E701" in lint-python, because the following annotation will cause a syntax error.
```python
a: int = 10
```
We can also leave E701 unchanged and just remove the relevant annotation. Do you have a better way? @hkvision 
Related link: https://github.com/intel-analytics/BigDL/pull/6829/files#diff-faf897596e92ba636c29683dd7988f3fead42125fbc7821b0e59d8e8ad297dddR62

**Comment by hkvision** [comment]
I don't think we need to type hint to class variables? Only arguments are enough?

**Comment by liangs6212** [comment]
> I don't think we need to type hint to class variables? Only arguments are enough?

Ok, I have removed some unnecessary annotation.

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 6]
```diff
@@ -124,21 +132,25 @@ def from_keras(*,
...
-            ):
+            data: Union[Callable, Tuple["ndarray", "ndarray"], "DataFrame"],
+            epochs: int=1,
+            validation_data: Optional[
+                Union[Callable, Tuple["ndarray", "ndarray"], "DataFrame"]]=None,
+            metric: Optional[Union[Callable, str]]=None,
+            metric_mode: Optional[str]=None,
+            metric_threshold: Optional[Union["Function", "Float", "Integer"]]=None,
+            n_sampling: int=1,
+            search_space: Optional[Dict[str, Union["Float", "Categorical", "Function", "Integer",
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:143]: Just a Optional[Dict] for covering more cases?

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 7]
```diff
@@ -124,21 +132,25 @@ def from_keras(*,
...
+                                                   Dict[str, List[bool]]]]]=None,
+            search_alg: Optional[str]=None,
+            search_alg_params: Optional[
+                Dict[str, Dict[str, Union["Function", "Integer", "Float"]]]]=None,
+            scheduler: Optional[str]=None,
+            scheduler_params: Optional[
+                Dict[str, Dict[str, Union["Function", "Integer", "Float"]]]]=None,
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:150]: same as above **if needed**

File: python/orca/src/bigdl/orca/automl/auto_estimator.py [hunk 8]
```diff
@@ -124,21 +132,25 @@ def from_keras(*,
...
+            validation_data: Optional[
+                Union[Callable, Tuple["ndarray", "ndarray"], "DataFrame"]]=None,
+            metric: Optional[Union[Callable, str]]=None,
+            metric_mode: Optional[str]=None,
+            metric_threshold: Optional[Union["Function", "Float", "Integer"]]=None,
+            n_sampling: int=1,
+            search_space: Optional[Dict[str, Union["Float", "Categorical", "Function", "Integer",
+                                                   Dict[str, List[bool]]]]]=None,
+            search_alg: Optional[str]=None,
+            search_alg_params: Optional[
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/auto_estimator.py:146]: same as above **if needed**

File: python/orca/src/bigdl/orca/automl/metrics.py [hunk 9]
```diff
@@ -21,12 +21,18 @@
...
 
 
 EPSILON = 1e-10
 
 
-def _standardize_input(y_true, y_pred, multioutput):
+def _standardize_input(y_true: "ndarray", y_pred: "ndarray",
+                       multioutput: str) -> Union[Tuple["ndarray", "ndarray", Tuple[int, int]],
+                                                  Tuple["ndarray", "ndarray", Tuple[()]],
+                                                  Tuple["ndarray", "ndarray", Tuple[int]]]:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/automl/metrics.py:35]: Just Tuple?

File: python/orca/src/bigdl/orca/automl/auto_estimator.py (hunk 6 above)
Comment (liangs6212) [reply python/orca/src/bigdl/orca/automl/auto_estimator.py]: Ok, It seems that Optional[Dict] is more concise.

**Comment by leonardozcm** [comment]
LGTM

**Pull request merged**